src/clout/core.cljc

```diff
@@ -0,0 +1,150 @@
+
+(defrecord CompiledRoute [source re keys absolute?]
+  Route
+  (route-matches [_ request]
+    (let [path-info (if absolute?
+                      (request-url request)
+                      (path-info request))]
+      #?(:clj
+         (let [matcher (re-matcher re path-info)]
+           (if (.matches matcher)
+             (assoc-keys-with-groups (re-groups* matcher) keys)))
+         :cljs
+         (if (.test re path-info)
+           (assoc-keys-with-groups (re-groups* re path-info) keys)))))
+  #?@(:clj [Object
+            (toString [_] source)]))
+
+(def ^:private route-parser
+  (insta/parser
+    #?(:clj
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^\\p{L}_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([\\p{L}_][\\p{L}_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'"
+       :cljs
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^A-Za-z_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([A-Za-z_][A-Za-z_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'")
+    :no-slurp true))
+
+(defn- parse [parser text]
+  (let [result (insta/parse parser text)]
+    (if (insta/failure? result)
+      (throw (ex-info "Parse error in route string" {:failure result}))
+      result)))
+
+(defn- find-route-key [form]
+  (case (first form)
+    :wildcard :*
+    :param (-> form second second keyword)))
+
+(defn- route-keys [parse-tree]
+  (->> (rest parse-tree)
+       (filter (comp #{:param :wildcard} first))
+       (map find-route-key)))
+
+(defn- trim-pattern [pattern]
+  (some-> pattern (subs 1 (dec (count pattern)))))
+
+(defn- param-regex [regexs key & [pattern]]
+  (str "(" (or (trim-pattern pattern) (regexs key) "[^/,;?]+") ")"))
+
+(defn- route-regex [parse-tree regexs]
+  (insta/transform
+    {:route    (comp re-pattern #?(:cljs #(str "^" % "$")) str)
+     :scheme   #(if (= % "//") "https?://" %)
+     :literal  re-escape
+     :escaped  #(re-escape (subs % 1))
+     :wildcard (constantly "(.*?)")
+     :param    (partial param-regex regexs)
+     :key      keyword
+     :pattern  str}
+    parse-tree))
+
+(defn- absolute-url? [path]
+  (boolean (re-matches #"(https?:)?//.*" path)))
+
+(defn route-compile
+  "Compile a route string for more efficient route matching."
+  ([path]
+   (route-compile path {}))
+  ([path regexs]
+   (let [ast (parse route-parser path)
+         ks  (route-keys ast)]
+     (assert (set/subset? (set (keys regexs)) (set ks))
+             "unused keys in regular expression map")
+     (CompiledRoute.
+       path
```

Comment:
This code looks incorrect indented as well. Can you follow the [clojure style guide](https://github.com/bbatsov/clojure-style-guide)?

Reply:
I'm trying... I relied on Cursive editor. It does code formatting quite well, but as I figured out, it does not provide facilities to configure ['single space indentation'](https://github.com/bbatsov/clojure-style-guide#one-space-indent) case.

Reply:
BTW if you're using Cursive, I believe you need to check the "One space list indent" option for Cursive to handle this automatically. I'm not sure why it's not the default.

![screen shot 2018-03-22 at 14 28 43](https://user-images.githubusercontent.com/8780/37776924-4dd9d6ac-2dde-11e8-81e4-b6dd7e22e91c.png)